wgengine/monitor/monitor_windows.go

```diff
@@ -7,273 +7,113 @@ package monitor
+			err = err1
 		}
+		m.routeChangeCallback = nil
 	}
+
+	return err
 }
 
-func (m *winMon) getIPOrRouteChangeMessage() (message, error) {
+func (m *winMon) Receive() (message, error) {
 	if m.ctx.Err() != nil {
+		m.logf("Receive call on closed monitor")
 		return nil, errClosed
 	}
 
-	// TODO(bradfitz): locking ourselves to an OS thread here
-	// likely isn't necessary, but also can't really hurt.
-	// We'll be blocked in windows.WaitForMultipleObjects below
-	// anyway, so might as well stay on this thread during the
-	// notify calls and cancel funcs.
-	// Given the past memory corruption from misuse of these APIs,
-	// and my continued lack of understanding of Windows APIs,
-	// I'll be paranoid. But perhaps we can remove this once
-	// we understand more.
-	runtime.LockOSThread()
-	defer runtime.UnlockOSThread()
-
-	addrHandle, oaddr, cancel, err := notifyAddrChange()
-	if err != nil {
-		m.logf("notifyAddrChange: %v", err)
-		return nil, err
-	}
-	defer cancel()
-
-	routeHandle, oroute, cancel, err := notifyRouteChange()
-	if err != nil {
-		m.logf("notifyRouteChange: %v", err)
-		return nil, err
-	}
-	defer cancel()
-
 	t0 := time.Now()
-	eventNum, err := windows.WaitForMultipleObjects([]windows.Handle{
-		m.closeHandle, // eventNum 0
-		addrHandle,    // eventNum 1
-		routeHandle,   // eventNum 2
-	}, false, windows.INFINITE)
-	if m.ctx.Err() != nil || (err == nil && eventNum == 0) {
-		return nil, errClosed
-	}
-	if err != nil {
-		m.logf("waitForMultipleObjects: %v", err)
-		return nil, err
-	}
-
-	d := time.Since(t0)
-	var eventStr string
 
-	// notifyAddrChange and notifyRouteChange both seem to return the same
-	// handle value. Determine which fired by looking at the "Internal" (sic)
-	// field of the Ovelapped instead.
-	// TODO(bradfitz): maybe clean this up; see TODO in callNotifyProc.
-	if (eventNum == 1 || eventNum == 2) && addrHandle == routeHandle {
-		if oaddr.Internal == _STATUS_WAIT_0 && oroute.Internal == _STATUS_PENDING {
-			eventStr = "addr-o" // "-o" overlapped suffix to distinguish from "addr" below
-		} else if oroute.Internal == _STATUS_WAIT_0 && oaddr.Internal == _STATUS_PENDING {
-			eventStr = "route-o"
-		} else {
-			eventStr = fmt.Sprintf("[unexpected] addr.internal=%d; route.internal=%d", oaddr.Internal, oroute.Internal)
-		}
-	} else {
-		switch eventNum {
-		case 1:
-			eventStr = "addr"
-		case 2:
-			eventStr = "route"
-		default:
-			eventStr = fmt.Sprintf("%d [unexpected]", eventNum)
-		}
-	}
-	m.logf("got windows change event after %v: evt=%s", d, eventStr)
-
-	m.mu.Lock()
-	{
-		m.lastNetChange = time.Now()
-
-		// Something changed, so assume Windows is about to
-		// discover its new proxy settings from WPAD, which
-		// seems to take a bit. Poll heavily for awhile.
-		m.inFastPoll = true
-		m.pollTicker.Reset(pollIntervalFast)
+	select {
+	case msg := <-m.messagec:
+		m.logf("got windows change event after %v: evt=%s", time.Since(t0).Round(time.Millisecond), msg.eventType)
```

Comment:
Yes, it can be done either way - how ever you decide. I've implemented it this way because it was the thing that the original implementation was logging. But I can change if you decide.